jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/DeltaSteppingShortestPath.java

```diff
@@ -0,0 +1,384 @@
+ * has high parallelism since all edges can be relaxed in parallel, the delta-stepping
+ * introduces parameter delta, which, when chooses optimally, yields still good parallelism
+ * and at the same time enables avoiding too many re-relaxations of the edges.
+ *
+ * <p>
+ * To prevent the necessity to synchronize threads the bucket structure is implemented here
+ * as a map of vertices to their bucket indices. Furthermore, every time a vertex is inserted
+ * in a bucket the tentative distance to the vertex and its predecessor should be updated in
+ * order to reconstruct the shortest paths tree at the end of the computation. Therefore to be
+ * able to update all the information safely it is kept in a single object which is wrapped
+ * in {@link AtomicReference}.
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ * @author Semen Chudakov
+ * @see AtomicReference
+ * @since August 2018
+ */
+public class DeltaSteppingShortestPath<V, E> extends BaseShortestPathAlgorithm<V, E> {
+    /**
+     * Error message for reporting the existence of an edge with negative weight.
+     */
+    private static final String NEGATIVE_EDGE_WEIGHT_NOT_ALLOWED = "Negative edge weight not allowed";
+    /**
+     * Error message for reporting that delta must be positive.
+     */
+    private static final String DELTA_MUST_BE_POSITIVE = "Delta must be positive";
+
+    /**
+     * The bucket width. A bucket with index $i$ therefore stores
+     * a vertex v if and only if v is queued and tentative distance
+     * to v $\in[i\cdot\Delta,(i+1)\cdot\Delta]$
+     */
+    private double delta;
+    /**
+     * Num of buckets in the bucket structure.
+     */
+    private int numOfBuckets;
+    /**
+     * Map with light edges for each vertex. An edge is considered
+     * light if its weight is less than or equal to {@link #delta}.
+     */
+    private Map<V, Set<E>> light;
+    /**
+     * Map with heavy edges for each vertex. An edge is
+     * considered heavy if its weight is greater than {@link #delta}.
+     */
+    private Map<V, Set<E>> heavy;
+
+    /**
+     * Map that stores information about each vertex.
+     *
+     * <p>
+     * In each triple the first value stands for the bucket index of a
+     * vertex or $-1$ if a vertex does not belong to any bucket. The second
+     * value stands for the tentative distance to a vertex. The third value
+     * of each triple stands for the predecessor of a vertex in the the
+     * shortest path tree. The second and the third values of each triple will
+     * be used at the end of the computation to construct shortest paths tree.
+     *
+     * <p>
+     * Keeping vertex information in an {@link AtomicReference} objects allows
+     * to avoid threads synchronisation. Thus a thread can safely update
+     * the information using standard CAS function.
+     */
+    private Map<V, AtomicReference<Triple<Integer, Double, E>>> verticesDataMap;
```

Comment:
I am not really sure whether this representation is enough in order to get good performance. You are filtering all vertices in order to find the vertices inside some bucket. I guess that you need to explicitly maintain the buckets.